source/common/router/metadatamatchcriteria_impl.cc

```diff
@@ -38,5 +38,21 @@ MetadataMatchCriteriaImpl::extractMetadataMatchCriteria(const MetadataMatchCrite
 
   return v;
 }
+
+MetadataMatchCriteriaConstPtr
+MetadataMatchCriteriaImpl::filterMatchCriteria(const std::set<std::string>& names) const {
+
+  std::vector<MetadataMatchCriterionConstSharedPtr> v;
+
+  // iterating over metadata_match_criteria_ ensures correct order without sorting
+  for (const auto& it : metadata_match_criteria_) {
+    if (names.count(it->name()) == 1) {
+      v.emplace_back(it);
+    }
+  }
+
+  return MetadataMatchCriteriaImplConstPtr(new MetadataMatchCriteriaImpl(v));
```

Comment:
Right, I think returning nullptr will be more consistent with code in other places and may save some memory. Changed.

I cannot use std::make_unique, because the constructor is private. Should I change it to public for that reason only or stick with current approach?

Reply:
Ah I see. In that case I think this is fine.